src/Forcings/discrete_forcing.jl

```diff
@@ -40,10 +40,12 @@ constraints on `typeof(parameters)`.
 """
 DiscreteForcing(func; parameters=nothing) = DiscreteForcing(func, parameters)
 
-@inline (forcing::DiscreteForcing)(i, j, k, grid, clock, model_fields) = 
-    forcing.func(i, j, k, grid, clock, model_fields, forcing.parameters)
+@inline function (forcing::DiscreteForcing{P, F})(i, j, k, grid, clock, model_fields) where {P, F<:Function}
+    parameters = forcing.parameters
```

Comment:
@ali-ramadhan can you see if these tests fix the "non-generic address space" LLVM error that you saw?

Reply:
Looks like this has been fixed on Buildkite as GPU forcing tests (part of model tests) all pass now: https://buildkite.com/clima/oceananigans/builds/49#9718f9bf-5859-4ce7-953d-1a0e6df53af3

You can see it fails with `LLVM error: Cannot cast between two non-generic address spaces` on master: https://buildkite.com/clima/oceananigans/builds/47#8970f9e4-4c09-47c0-97f2-e413cfc38cee/31-221

Reply:
Nice! There are two distinct changes here: first, `getproperty(forcing, :parameters)` was moved outside the function signature. Then we also force specialization on the type of `forcing.func`. I don't know which was helpful (maybe both), or, more importantly, _why_...